frame/evm/precompile/modexp/src/lib.rs

```diff
@@ -77,15 +78,19 @@ fn calculate_gas_cost(
 			//   must be > 0)
 			// * the addition can't overflow because the terms are both capped at roughly
 			//   8 * max size of exp_length (1024)
-			iteration_count =
-				(8 * (exp_length - 32)) + exponent.bitand(max_256_bit_uint).bits() - 1;
+			// * the EIP spec is written in python, in which (exponent & (2**256 - 1)) takes the
+			//   FIRST 32 bytes. However this `BigUint` `&` operator takes the LAST 32 bytes.
+			//   We thus instead take the bytes manually.
+			let exponent_head = BigUint::from_bytes_be(&exponent_bytes[..32]);
```

Comment:
I don't see that there is any guarantee on the length of `exponent_bytes`. It looks like its length comes from user input, which can be between `0..1024`, right?

Reply:
It's defined below as:

```rust
let mut exp_buf = vec![0u8; exp_len];
```

where `exp_len` is user input